Add the "gs:" option to filter strings in the lobby

GenjisRevenge opened the pull request:
This feature allows the player to select game servers in the lobby with filter strings like `gs:1,2 baal`.

I haven't tested this change because the README.md of this repo advises against it:

> Any debug or changed version will only work in Single Player. Do not enter multiplayer with a modified BH

The bulk of the logic (the `ParsedFilterString` class) was implemented in a small standalone project where I tested it manually so that shouldn't be extremely buggy. You have to make sure that the untested changes (in the `GameFilter` class) work as intended when the game goes online and enters the lobby.

croboy commented:
Wow this is awesome

Canite commented:
I'll look into this soon and check it out, haven't had time to go through it.

fergushev commented:
Hey @GenjisRevenge, I don't know if you've been keeping up with the dev streams but we've added difficulty filter buttons to the join game screen. As part of this, the underlying behavior of the "GS" text in the game list has also changed. What this means for this PR:

- The difficulty filter part of your change wont be needed anymore and can be removed
- The GS filter could still be a useful addition, but would need to be modified as the GS information is no longer part of the game description. It's now part of the gamelist packet: https://github.com/Project-Diablo-2/BH/blob/main/BH/Modules/Gamefilter/Gamefilter.cpp#L85

If you have any questions please let me know. Happy to help!

GenjisRevenge commented:
Hi @fergushev

I updated the branch: removed the difficulty filter and getting the game server number from the new location.
However, I didn't try to log in with these modifications to the live server. You have to test run it before deployment.

fergushev commented:
@GenjisRevenge Looks good to me. Thanks for the update!
